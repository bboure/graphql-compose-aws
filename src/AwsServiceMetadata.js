/* @flow */

import { schemaComposer, ObjectTypeComposer, upperFirst } from 'graphql-compose';
import type { GraphQLFieldConfig } from 'graphql-compose/lib/graphql';

export type ServiceMetadataConfig = {
  apiVersion: string,
  endpointPrefix: string,
  globalEndpoint: string,
  serviceAbbreviation: string,
  serviceFullName: string,
  signatureVersion: string,
  uid: string,
};

export class AwsServiceMetadata<TContext> {
  metadata: ServiceMetadataConfig;
  tc: ?ObjectTypeComposer<any, TContext>;

  constructor(metadata: ServiceMetadataConfig) {
    this.metadata = metadata;
  }

  getPrefix(): string {
    return `Aws${upperFirst(this.metadata.endpointPrefix)}`;
  }

  getDescription(): string {
    // { apiVersion: '2006-03-01',
    //   checksumFormat: 'md5',
    //   endpointPrefix: 's3',
    //
    //   globalEndpoint: 's3.amazonaws.com',
    //   protocol: 'rest-xml',
    //   serviceAbbreviation: 'Amazon S3',
    //   serviceFullName: 'Amazon Simple Storage Service',
    //   serviceId: 'S3',
    //   signatureVersion: 's3',
    //   timestampFormat: 'rfc822',
    //   uid: 's3-2006-03-01' }
    const { serviceFullName, apiVersion } = this.metadata;
    return `${serviceFullName} (${apiVersion})`;
  }

  getTypeComposer(): ObjectTypeComposer<any, TContext> {
    if (!this.tc) {
      this.tc = schemaComposer.createObjectTC(`
        type ${this.getPrefix()}Metadata {
          apiVersion: String
          endpointPrefix: String
          globalEndpoint: String
          serviceAbbreviation: String
          serviceFullName: String
          signatureVersion: String
          uid: String
          raw: JSON
        }
      `);
    }
    return this.tc;
  }

  getFieldConfig(): GraphQLFieldConfig<any, any> {
    return {
      type: this.getTypeComposer().getType(),
      resolve: () => {
        return {
          ...this.metadata,
          raw: this.metadata,
        };
      },
    };
  }
}
