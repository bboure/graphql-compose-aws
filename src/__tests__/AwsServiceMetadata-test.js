/* @flow */

import { ObjectTypeComposer } from 'graphql-compose';
import { AwsServiceMetadata } from '../AwsServiceMetadata';
import s3Cfg from '../__mocks__/s3-2006-03-01.json';

const meta = new AwsServiceMetadata(s3Cfg.metadata);

describe('AwsServiceMetadata', () => {
  it('getPrefix()', () => {
    expect(meta.getPrefix()).toBe('AwsS3');
  });

  it('getTypeComposer()', () => {
    const tc = meta.getTypeComposer();
    expect(tc).toBeInstanceOf(ObjectTypeComposer);
    expect(tc.getTypeName()).toBe('AwsS3Metadata');
    expect(tc.getFieldNames()).toMatchSnapshot();
  });

  it('getFieldConfig()', () => {
    const fc: any = meta.getFieldConfig();
    expect(fc).toMatchSnapshot();
    expect(fc.resolve()).toMatchSnapshot();
  });
});
