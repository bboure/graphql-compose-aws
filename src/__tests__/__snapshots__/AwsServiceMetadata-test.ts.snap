// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AwsServiceMetadata getFieldConfig() 1`] = `
Object {
  "resolve": [Function],
  "type": "AwsS3Metadata",
}
`;

exports[`AwsServiceMetadata getFieldConfig() 2`] = `
Object {
  "apiVersion": "2006-03-01",
  "checksumFormat": "md5",
  "endpointPrefix": "s3",
  "globalEndpoint": "s3.amazonaws.com",
  "protocol": "rest-xml",
  "raw": Object {
    "apiVersion": "2006-03-01",
    "checksumFormat": "md5",
    "endpointPrefix": "s3",
    "globalEndpoint": "s3.amazonaws.com",
    "protocol": "rest-xml",
    "serviceAbbreviation": "Amazon S3",
    "serviceFullName": "Amazon Simple Storage Service",
    "serviceId": "S3",
    "signatureVersion": "s3",
    "timestampFormat": "rfc822",
    "uid": "s3-2006-03-01",
  },
  "serviceAbbreviation": "Amazon S3",
  "serviceFullName": "Amazon Simple Storage Service",
  "serviceId": "S3",
  "signatureVersion": "s3",
  "timestampFormat": "rfc822",
  "uid": "s3-2006-03-01",
}
`;

exports[`AwsServiceMetadata getTypeComposer() 1`] = `
Array [
  "apiVersion",
  "endpointPrefix",
  "globalEndpoint",
  "serviceAbbreviation",
  "serviceFullName",
  "signatureVersion",
  "uid",
  "raw",
]
`;
